eventsources/sources/azureservicebus/start.go

```diff
@@ -0,0 +1,182 @@
+			ReceiveMode: servicebus.ReceiveModeReceiveAndDelete,
+		})
+	}
+	if err != nil {
+		if receiverType == ReceiverTypeQueue {
+			log.With("queue", servicebusEventSource.QueueName).Errorw("failed to create a queue receiver", zap.Error(err))
+		} else {
+			log.With("topic", servicebusEventSource.TopicName, "subscription", servicebusEventSource.SubscriptionName).Errorw("failed to create a receiver for subscription", zap.Error(err))
+		}
+		return err
+	}
+
+	if servicebusEventSource.JSONBody {
+		log.Info("assuming all events have a json body...")
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("context has been cancelled, stopping the Azure Service Bus event source...")
+			if err := receiver.Close(ctx); err != nil {
+				log.Errorw("failed to close the receiver", zap.Error(err))
+				return err
+			}
+			return nil
+		default:
+			messages, err := receiver.ReceiveMessages(ctx, 1, nil)
+			if err != nil {
+				log.Errorw("failed to receive messages", zap.Error(err))
+				return err
+			}
+
+			for _, message := range messages {
+				if err := el.handleOne(servicebusEventSource, message, dispatch, log); err != nil {
+					if receiverType == ReceiverTypeQueue {
+						log.With("queue", servicebusEventSource.QueueName, "message_id", message.MessageID).Errorw("failed to process Azure Service Bus message", zap.Error(err))
+					} else {
+						log.With("topic", servicebusEventSource.TopicName, "subscription", servicebusEventSource.SubscriptionName, "message_id", message.MessageID).Errorw("failed to process Azure Service Bus message", zap.Error(err))
+					}
+					el.Metrics.EventProcessingFailed(el.GetEventSourceName(), el.GetEventName())
+					continue
+				}
+			}
+		}
+	}
+}
+
+func (el *EventListener) handleOne(servicebusEventSource *v1alpha1.AzureServiceBusEventSource, message *servicebus.ReceivedMessage, dispatch func([]byte, ...eventsourcecommon.Option) error, log *zap.SugaredLogger) error {
+	defer func(start time.Time) {
+		el.Metrics.EventProcessingDuration(el.GetEventSourceName(), el.GetEventName(), float64(time.Since(start)/time.Millisecond))
+	}(time.Now())
+
+	log.Infow("received the message", zap.Any("message_id", message.MessageID))
+	eventData := &events.AzureServiceBusEventData{
+		ApplicationProperties: message.ApplicationProperties,
+		ContentType:           message.ContentType,
+		CorrelationId:         message.CorrelationID,
+		EnqueuedTime:          message.EnqueuedTime,
+		MessageId:             message.MessageID,
+		ReplyTo:               message.ReplyTo,
+		SequenceNumber:        message.SequenceNumber,
+		Subject:               message.Subject,
+		Metadata:              servicebusEventSource.Metadata,
+	}
+
+	if servicebusEventSource.JSONBody {
+		eventData.Body = (*json.RawMessage)(&message.Body)
+	} else {
+		eventData.Body = message.Body
+	}
+
+	eventBytes, err := json.Marshal(eventData)
+	if err != nil {
+		el.Metrics.EventProcessingFailed(el.GetEventSourceName(), el.GetEventName())
```

Comment:
Will this cause duplicate counting since there's already one in line 135?

Reply:
You are correct, I will remove it.